core/Decode.cpp

```diff
@@ -148,20 +173,151 @@ namespace olympia
             // Send instructions on their way to rename
             for (uint32_t i = 0; i < num_decode; ++i)
             {
-                const auto & inst = fetch_queue_.read(0);
-                insts->emplace_back(inst);
-                inst->setStatus(Inst::Status::DECODED);
-
-                if (fusion_enable_)
+                if (uop_queue_.size() > 0)
                 {
-                    uids.push_back(inst->getMavisUid());
+                    const auto & inst = uop_queue_.read(0);
+                    insts->emplace_back(inst);
+                    inst->setStatus(Inst::Status::DECODED);
+                    ILOG("From UOp Queue Decoded: " << inst);
+                    uop_queue_.pop();
+                }
+                else
+                {
+                    auto & inst = fetch_queue_.read(0);
+
+                    // for vector instructions, we block on vset and do not allow any other
+                    // processing of instructions until the vset is resolved optimizations could be
+                    // to allow scalar operations to move forward until a subsequent vector
+                    // instruction is detected or do vset prediction
+
+                    // mavis_id 214 -> vsetvl
+                    // mavis_id 216 -> vsetivli
+                    // TODO: change back from getMnemonic() to getInstructionUniqueID() where the ID
+                    // doesn't change
+                    if (inst->getOpCodeInfo()->getInstructionUniqueID() == mavis_vsetivli_uid_)
+                    {
+                        // vsetivli with immediates, we can set at decode and continue to process
+                        // instruction group, no vset stall
+                        VCSRs_.lmul = inst->getLMUL();
```

Comment:
How about a method to the `VCSRs_` struct called `update` that takes an instruction and updates all of these values? Would reduced code duplication here and in the `process_vset_` method.

Reply:
These if-statements need to be restructured. I don't think the `vsetvli` when the source is not x0 case is handled. Something like this should work:
```
if(vset) {
    if(non-stalling vset) {
        // process vset
    }
    else {
        // handle stall
    }
}
elseif (vector inst) {
    // uop gen
}

// send to decode
```

Reply:
Fixed in [73ca9f4](https://github.com/riscv-software-src/riscv-perf-model/pull/167/commits/73ca9f4adb0bce0b8756a9b1ab3d88bf96c589bd)